### ISSUE TITLE
Proposal for handling Software Collections

### ISSUE THREAD

**Issue by bkabrda** [sosreport/sos#898]
Hi, I'd like to add support for handling Software Collections [1] into sosreport. Since I think this is not an easy issue, I wanted to send a proposal first to see what you think about it. I'll work on the implementation when we agree on what the support should look like.

Basically, I see 2 different things we might want to do here:
* Listing all software collections on a system
* Running various commands for all (or subset of) individual collections

The first case is easy, one just needs to create a simple plugin that will run `scl list-collections` (or alternatively `scl -l` for older releases of scl-utils).

The second part will be much harder. The tricky part is, that I think we don't want to get limited to a hardcoded set of SCLs and having a plugin for each separate collection. Therefore I think it should be possible to have a plugin that will be run for any collection that fulfils certain criteria. I think the best approach would be to
* create a new `Plugin` subclass `SCLPlugin`
* allow using `%(scl_name)` in `class.packages` and `class.files` for subclasses of `SCLPlugin`
* creating and running a plugin instance for every SCL for which the conditions are met (which means after doing the interpolation of `class.packages` and `class.files` strings and checking these conditions the standard way)

For example, if I wanted to do something for every SCL that contained `npm` package, I could do:

```
class NPMSCLCommand(Plugin, SCLPlugin):
    packages = ("%(scl_name)-npm",)

    def setup(self):
        pass  # do something useful for every SCL that contains `npm` package
```

Furthermore, the `SCLPlugin` constructor would probably accept `scl_name` as an argument, so that it would be accessible via `self.scl_name` during `setup`.

Please let me know if this makes sense to you, or if you think modifications are needed.
Thanks!

[1] https://www.softwarecollections.org/en/

**Comment by bmr-cymru** [comment]
From a quick scan this seems reasonable - it's been on my TODO for a long time, so thanks for working on this.

I'd also reached the conclusion that modifying a `Plugin` class was necessary for reasonably elegant SCL support; the one thing that needs some consideration to get right here is the interaction between this new `Plugin`subclass and the "tagging class" mechanism that is used to mark plugins as compatible with particular policy families.

> Furthermore, the SCLPlugin constructor would probably accept scl_name 

This is somewhat problematic in that it will be inconsistent with constructors for all other plugins. This could be tricky since plugins are instantiated by the `load_plugins()` loop in `sosreport.py`; from a first consideration I'd rather not start exposing bits of SCL logic there - conventionally this sort of system detail is handled in the policy classes. Another issue is that plugin constructors are often chained between tagging class subclasses and the parent class - it's not a big deal since plugin authors will need to be aware if they are subclassing an `SCLPlugin` anyway but another possible detail to deal with or document.

I'll have a more detailed think about it over the weekend but I think this can be made into something workable.

**Comment by bkabrda** [comment]
Thanks a lot for your reaction!

I have yet to consider how to approach the "tagging class", since I'm not that familiar with the codebase.

As for the constructor accepting `scl_name`, I think we can skip that for now and only consider it if we find that it's really necessary.

Based on your comment, I think my proposal is reasonable enough for me to try to put together a working prototype and then I'll see what further questions/problems arise while doing that.

**Comment by bmr-cymru** [comment]
👍

**Comment by bmr-cymru** [comment]
I'm working through sos reviews and merges this week for the first time in a while - I have a few other small reviews that are higher up on the list, but I hope to be able to have a proper look at this one before the end of the week.

**Comment by bmr-cymru** [comment]
Moving out to 3.5.

**Comment by sandrobonazzola** [comment]
We're interested in this feature as well for oVirt since we are now using PostgreSQL95 from SoftwareCollections. I think @mwperina, @tiraboschi  and @didib may be interested.

**Comment by bmr-cymru** [comment]
Merged for 3.5.
